core/trino-spi/src/main/java/io/trino/spi/type/Decimals.java

```diff
@@ -287,6 +289,33 @@ public static BigDecimal rescale(BigDecimal value, DecimalType type)
         return value;
     }
 
+    public static long rescaleAndRoundHalfUp(long value, int fromScale, int toScale)
+    {
+        return shortToShortCast(
+                value,
+                MAX_SHORT_PRECISION,
+                fromScale,
+                MAX_SHORT_PRECISION,
+                toScale,
+                longTenToNth(abs(fromScale - toScale)),
+                longTenToNth(abs(fromScale - toScale)) / 2);
+    }
+
+    public static BigInteger rescaleAndRoundHalfUp(BigInteger value, int fromScale, int toScale)
+    {
+        if (toScale < fromScale) {
+            return decodeUnscaledValue(
```

Comment:
Why not to use native `BigInteger` capabilities instead of going though `BigInteger -> Slice -> BigInteger` conversion?

Reply:
Was not sure about the performance impact of that so decided to go with what was present in the code base. Will using `BigInteger` be better?

Reply:
Also, `BigInteger` does not have a round function. So the alternative was `BigInteger -> BigDecimal -> BigInteger`.